File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Tenant/TenantContext.cs
namespace JoeDayz.Microservicios.Modulo01.Tenant;

/// <summary>
/// Contexto de tenant de la peticion actual.
/// Equivalente .NET del <c>ThreadLocal</c> de Java: <see cref="AsyncLocal{T}"/> fluye
/// a traves de <c>await</c>, por lo que funciona igual con async/await y con Task.Run.
/// En ASP.NET Core esto se poblaria en un middleware a partir del subdominio o del JWT.
/// </summary>
public static class TenantContext
{
    private static readonly AsyncLocal<TenantId?> Current = new();

    public static void Set(TenantId tenantId) => Current.Value = tenantId;

    public static TenantId Require() =>
        Current.Value
        ?? throw new InvalidOperationException(
            "No hay tenant en el contexto: toda operacion multi-tenant requiere un TenantId");

    public static void Clear() => Current.Value = null;
}

File: modulo-05-persistencia-poliglota/notification-service-spring/src/main/java/pe/joedayz/microservicios/notification/tenant/TenantContext.java
package pe.joedayz.microservicios.notification.tenant;

public final class TenantContext {

    private static final ThreadLocal<String> CURRENT = new ThreadLocal<>();

    private TenantContext() {
    }

    public static void set(String tenantId) {
        CURRENT.set(tenantId);
    }

    public static String require() {
        String tenant = CURRENT.get();
        if (tenant == null || tenant.isBlank()) {
            throw new IllegalStateException("Header X-Tenant-ID requerido");
        }
        return tenant;
    }

    public static void clear() {
        CURRENT.remove();
    }
}
